akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java

```diff
@@ -0,0 +1,347 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.compress;
+
+import java.io.UnsupportedEncodingException;
+import java.util.Random;
+
+/**
+ * Count-Min Sketch datastructure.
+ * An Improved Data Stream Summary: The Count-Min Sketch and its Applications
+ * https://web.archive.org/web/20060907232042/http://www.eecs.harvard.edu/~michaelm/CS222/countmin.pdf
+ * 
+ * This implementation is mostly taken and slightly adjusted from the Apache V2 licensed project `stream-lib`, located here:
+ * https://github.com/clearspring/stream-lib/blob/master/src/main/java/com/clearspring/analytics/stream/frequency/CountMinSketch.java
+ */
+public class CountMinSketch {
+
+  public static final long PRIME_MODULUS = (1L << 31) - 1;
+
+  private int depth;
+  private int width;
+  private long[][] table;
+  private long[] hashA;
+  private long size;
+  private double eps;
+  private double confidence;
+
+  public CountMinSketch(int depth, int width, int seed) {
+    this.depth = depth;
+    this.width = width;
+    this.eps = 2.0 / width;
+    this.confidence = 1 - 1 / Math.pow(2, depth);
+    initTablesWith(depth, width, seed);
+  }
+
+  public CountMinSketch(double epsOfTotalCount, double confidence, int seed) {
+    // 2/w = eps ; w = 2/eps
+    // 1/2^depth <= 1-confidence ; depth >= -log2 (1-confidence)
+    this.eps = epsOfTotalCount;
+    this.confidence = confidence;
+    this.width = (int) Math.ceil(2 / epsOfTotalCount);
+    this.depth = (int) Math.ceil(-Math.log(1 - confidence) / Math.log(2));
+    initTablesWith(depth, width, seed);
+  }
+
+  CountMinSketch(int depth, int width, int size, long[] hashA, long[][] table) {
+    this.depth = depth;
+    this.width = width;
+    this.eps = 2.0 / width;
+    this.confidence = 1 - 1 / Math.pow(2, depth);
+    this.hashA = hashA;
+    this.table = table;
+    this.size = size;
+  }
+
+  private void initTablesWith(int depth, int width, int seed) {
+    this.table = new long[depth][width];
+    this.hashA = new long[depth];
+    Random r = new Random(seed);
+    // We're using a linear hash functions
+    // of the form (a*x+b) mod p.
```

Comment:
Just copy out the necessary code from Murmur3 impl?

On Tue, Jun 7, 2016 at 10:57 AM, Konrad Malawski <notifications@github.com>
wrote:

> In
> akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java
> <https://github.com/akka/akka/pull/20546#discussion_r66033001>:
>
> > +  CountMinSketch(int depth, int width, int size, long[] hashA, long[][] table) {
> > +    this.depth = depth;
> > +    this.width = width;
> > +    this.eps = 2.0 / width;
> > +    this.confidence = 1 - 1 / Math.pow(2, depth);
> > +    this.hashA = hashA;
> > +    this.table = table;
> > +    this.size = size;
> > +  }
> > +
> > +  private void initTablesWith(int depth, int width, int seed) {
> > +    this.table = new long[depth][width];
> > +    this.hashA = new long[depth];
> > +    Random r = new Random(seed);
> > +    // We're using a linear hash functions
> > +    // of the form (a*x+b) mod p.
>
> This is murmur2. I did not manage to move to Scalas murmur3 yet due to
> java scala interop issues. Its on the list to do. I'll write up.remaining
> things to be done like that. I first need to get it working then replace
> hashes etc
> … <#m_1377374578521158062_>
> On Jun 7, 2016 10:27, "drewhk" ***@***.***> wrote: In
> akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java <#20546
> (comment) <https://github.com/akka/akka/pull/20546#discussion_r66028767>>:
> > + CountMinSketch(int depth, int width, int size, long[] hashA, long[][]
> table) { > + this.depth = depth; > + this.width = width; > + this.eps = 2.0
> / width; > + this.confidence = 1 - 1 / Math.pow(2, depth); > + this.hashA =
> hashA; > + this.table = table; > + this.size = size; > + } > + > + private
> void initTablesWith(int depth, int width, int seed) { > + this.table = new
> long[depth][width]; > + this.hashA = new long[depth]; > + Random r = new
> Random(seed); > + // We're using a linear hash functions > + // of the form
> (a*x+b) mod p. The one included here is Murmur3? — You are receiving this
> because you authored the thread. Reply to this email directly, view it on
> GitHub <
> https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66028767>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/AAHYk1vSvMnaLhe8iC2S0xmZTbjMvWiJks5qJSuIgaJpZM4IhRoS>
> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66033001>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA45vfOLs8lTf5c8xLQjCpAmRKirudOrks5qJTJlgaJpZM4IhRoS>
> .
>